T4930: Allow WireGuard peers via DNS hostname

sskaje opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->

T4930: Allow WireGuard peers via DNS hostname + new script resetting peer
T4930: Ensure peer is created even if dns not working
T4930: limit wg retry times by using environment variable
T4930: make wg dns retry configurable through interfaces wireguard wgX max-dns-retry



## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T4930

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
wireguard

## Proposed changes
<!--- Describe your changes in detail -->

1. Support wireguard endpoint using domain as peer address, make sure `wg` retries for dns resolution no more than 5 times;
2. Introduce an op mode command `reset wireguard`, if user want to use `wg set` to force wireguard redo dns resolution;
3. Detail changes like splitting Wireguard peer creation into peer creation + endpoint setup, with exception (need help on correct exception/log handling, not found in docs.vyos.io)

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

config mode
```
vyos@vyos# set interfaces wireguard wg0 peer wg0-xxx address t2.vm.xxx.xxx
[edit]
vyos@vyos# commit
[edit]
vyos@vyos#
```

op mode

```
# reset all peers on wg0
vyos@vyos:~$ reset wireguard interface wg0
Resetting wg0 peer YYYY= endpoint to t1.vm.xxx.xxx:ppppp ... done
Resetting wg0 peer XXXX= endpoint to t2.vm.xxxx.xxx:ppppp ... done

# reset single peer on wg0
vyos@vyos:~$ reset wireguard interface wg0  peer wg0-xxx
Resetting wg0 peer XXXX= endpoint to t2.vm.xxxx.xxx:ppppp ... done

```

vyos without working dns

I've provided screenshots in task's comments.


configure max dns resolution retry times

```
vyos@vyos# set interfaces wireguard wg0
Possible completions:
+  address              IP address
   description          Description
   disable              Administratively disable interface
   domain-name          Endpoint Domain Name
   fwmark               A 32-bit fwmark value set on all outgoing packets (default: 0)
 > ip                   IPv4 routing parameters
 > ipv6                 IPv6 routing parameters
   max-dns-retry        Max retry when DNS resolves failed. (default: 3)
 > mirror               Mirror ingress/egress packets
   mtu                  Maximum Transmission Unit (MTU) (default: 1420)
+> peer                 peer alias
   per-client-thread    Process traffic from each client in a dedicated thread
   port                 Port number used by connection
   private-key          Base64 encoded private key
   redirect             Redirect incoming packet to destination
   vrf                  VRF instance name


[edit]
vyos@vyos# set interfaces wireguard wg0 max-dns-retry
Possible completions:
   <1-15>               Max retry times



[edit]
vyos@vyos# set interfaces wireguard wg0 max-dns-retry 3
[edit]
vyos@vyos#
```

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
```
root@vyos:/home/vyos# python3 /usr/libexec/vyos/tests/smoke/cli/test_interfaces_wireguard.py
test_01_wireguard_peer (__main__.WireGuardInterfaceTest.test_01_wireguard_peer) ... ok
test_02_wireguard_add_remove_peer (__main__.WireGuardInterfaceTest.test_02_wireguard_add_remove_peer) ... ok
test_03_wireguard_same_public_key (__main__.WireGuardInterfaceTest.test_03_wireguard_same_public_key) ... ok
test_04_wireguard_threaded (__main__.WireGuardInterfaceTest.test_04_wireguard_threaded) ... ok
test_05_wireguard_peer_pubkey_change (__main__.WireGuardInterfaceTest.test_05_wireguard_peer_pubkey_change) ... ok

----------------------------------------------------------------------
Ran 5 tests in 52.648s

OK
root@vyos:/home/vyos#
```


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [X] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [X] I have linked this PR to one or more Phabricator Task(s)
- [X] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [X] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

sever-sever commented:
Build package fails (based on CI)

```
PYTHONPATH=python/ python3 -m "nose" --with-xunit src --with-coverage --cover-erase --cover-xml --cover-package src/conf_mode,src/op_mode,src/completion,src/helpers,src/validators,src/tests --verbose
*** Error compiling './python/vyos/ifconfig/wireguard.py'...
  File "./python/vyos/ifconfig/wireguard.py", line 179
    f'Resetting {self.config['ifname']} peer {public_key} endpoint to {address}:{port} ... ',
                              ^^^^^^
SyntaxError: f-string: unmatched '['

make[2]: *** [Makefile:89: test] Error 1
make[2]: Leaving directory '/__w/vyos-1x/vyos-1x/packages/vyos-1x'
make[1]: *** [debian/rules:31: override_dh_auto_build] Error 2
make[1]: Leaving directory '/__w/vyos-1x/vyos-1x/packages/vyos-1x'
make: *** [debian/rules:21: build] Error 2
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
Error: Process completed with exit code
```

sskaje commented:
> Build package fails (based on CI)
> 
> ```
> PYTHONPATH=python/ python3 -m "nose" --with-xunit src --with-coverage --cover-erase --cover-xml --cover-package src/conf_mode,src/op_mode,src/completion,src/helpers,src/validators,src/tests --verbose
> *** Error compiling './python/vyos/ifconfig/wireguard.py'...
>   File "./python/vyos/ifconfig/wireguard.py", line 179
>     f'Resetting {self.config['ifname']} peer {public_key} endpoint to {address}:{port} ... ',
>                               ^^^^^^
> SyntaxError: f-string: unmatched '['
> 
> make[2]: *** [Makefile:89: test] Error 1
> make[2]: Leaving directory '/__w/vyos-1x/vyos-1x/packages/vyos-1x'
> make[1]: *** [debian/rules:31: override_dh_auto_build] Error 2
> make[1]: Leaving directory '/__w/vyos-1x/vyos-1x/packages/vyos-1x'
> make: *** [debian/rules:21: build] Error 2
> dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
> Error: Process completed with exit code
> ```

Updated.

My IDE was set python3.12, that syntax is acceptable in 3.12 but not in 3.11  XD

sarthurdev commented on op-mode-definitions/reset-wireguard.xml.in:
```diff
@@ -0,0 +1,48 @@
+              <help>Reset specific WireGuard interface</help>
+            </properties>
+            <children>
+              <tagNode name="">
```
The tag node should be named `interface` instead of the regular node and this empty tag node.

sarthurdev commented on op-mode-definitions/reset-wireguard.xml.in:
```diff
@@ -0,0 +1,48 @@
+                      <help>Reset specific WireGuard peer</help>
+                    </properties>
+                    <children>
+                      <tagNode name="">
```
The tag node should be named `peer` instead of the regular node and this empty tag node.

sarthurdev reviewed (changes requested):
Does the `WG_ENDPOINT_RESOLUTION_RETRIES` var block execution while waiting to resolve?

sever-sever commented on src/op_mode/reset_wireguard.py:
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2022-2023 VyOS maintainers and contributors
```
There should be 2024

sskaje commented:
> Does the `WG_ENDPOINT_RESOLUTION_RETRIES` var block execution while waiting to resolve?

Sorry, I don't understand your question.

WG_ENDPOINT_RESOLUTION_RETRIES is an environment variable use in `wg` command, source here:

`wg` reads env: 
https://git.zx2c4.com/wireguard-tools/tree/src/config.c#n180

`wg` assigns as local variable: 
https://git.zx2c4.com/wireguard-tools/tree/src/config.c#n199

`wg` loops and retries:
https://git.zx2c4.com/wireguard-tools/tree/src/config.c#n259

By default, `wg` retries 15 times, delay time increased to 1.2x of last delay time on each failure.
This make `wg` keeps wait for around 1 minutes until dns resolved or all attempts failed.


BTW, the `reset-wireguard.xml.in` is simplified by following your tips, please check if there is anything else I didn't make right. 
That was AI generated code, because my earlier attempts on completions were failed.

sarthurdev commented:
> By default, `wg` retries 15 times, delay time increased to 1.2x of last delay time on each failure.
> This make `wg` keeps wait for around 1 minutes until dns resolved or all attempts failed.

That's what I was concerned by, we want to avoid long boot/commit time because of DNS resolution not being available.

sskaje commented:
> > By default, `wg` retries 15 times, delay time increased to 1.2x of last delay time on each failure.
> > This make `wg` keeps wait for around 1 minutes until dns resolved or all attempts failed.
> 
> That's what I was concerned by, we want to avoid long boot/commit time because of DNS resolution not being available.

Yes, that's why I set it to 5 and I still feel it costs too much time. 

How about 3x? Or 3x by default and let user can customize it in somewhere like `set interfaces wireguard wg0 max-dns-retry`?

sskaje commented:
`max-dns-retry` added, limit 1-15.

```
vyos@vyos# set interfaces wireguard wg0
Possible completions:
+  address              IP address
   description          Description
   disable              Administratively disable interface
   domain-name          Endpoint Domain Name
   fwmark               A 32-bit fwmark value set on all outgoing packets (default: 0)
 > ip                   IPv4 routing parameters
 > ipv6                 IPv6 routing parameters
   max-dns-retry        Max retry when DNS resolves failed. (default: 3)
 > mirror               Mirror ingress/egress packets
   mtu                  Maximum Transmission Unit (MTU) (default: 1420)
+> peer                 peer alias
   per-client-thread    Process traffic from each client in a dedicated thread
   port                 Port number used by connection
   private-key          Base64 encoded private key
   redirect             Redirect incoming packet to destination
   vrf                  VRF instance name


[edit]
vyos@vyos# set interfaces wireguard wg0 max-dns-retry
Possible completions:
   <1-15>               Max retry times



[edit]
vyos@vyos# set interfaces wireguard wg0 max-dns-retry 3
[edit]
vyos@vyos#
```

sskaje commented:
Hi reviewers, 

Still need help:

1. Exception handling in WireGuardIf.update(). print() or any best practice in vyos?
2. vyos-domain-resolver part in interfaces_wireguard, need help about the logic: as I read what's in `nat` & `firewall`, the service will keep running if ip_fqdn or ip6_fqdn not empty. That means I should check all working peers and build a similar object storing peers with domain, and check if this object not empty to create the /run/use-vyos-domain-resolver-*?

dmbaturin commented:
The current design seems promising, we'll get back to you with comments about implementation if we find anything that we think we can improve.

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -24,7 +24,7 @@
 from vyos.ifconfig import Interface
 from vyos.ifconfig import Operational
-from vyos.template import is_ipv6
+from vyos.template import is_ipv6, is_ipv4
```
Please move this to a new line - it makes search/replace with patterns more easy.

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -156,6 +156,63 @@ def show_interface(self):
         return answer
 
+    def get_latest_handshakes(self):
+
```
Remove stray newline

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -156,6 +156,63 @@ def show_interface(self):
+        output = {}
+
+        # Dump wireguard last handshake
+        _f = self._cmd(f'wg show {self.config["ifname"]} latest-handshakes')
```
```suggestion
        _f = self._cmd(f'wg show {self.ifname} latest-handshakes')
```

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -156,6 +156,63 @@ def show_interface(self):
+    def reset_peer(self, peer_name=None, public_key=None):
+        from vyos.config import Config
+
+        c = Config()
```
Please use this isntead:

```python
from vyos.configquery import ConfigTreeQuery
config = ConfigTreeQuery()
```

It's designed for op-mode.

`return_effective_value` becomes `value` ... just see the implementation of this lightweight class.
`list_effective_nodes` becomes `list_nodes`

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -156,6 +156,63 @@ def show_interface(self):
+                if c.exists_effective(['peer', peer, 'disable']):
+                    continue
+
+                cmd = f"wg set {self.config['ifname']} peer {peer_public_key} endpoint {address}:{port}"
```
```suggestion
                cmd = f"wg set {self.ifname} peer {peer_public_key} endpoint {address}:{port}"
```

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -156,6 +156,63 @@ def show_interface(self):
+                cmd = f"wg set {self.config['ifname']} peer {peer_public_key} endpoint {address}:{port}"
+                try:
+                    print(
+                        f'Resetting {self.config["ifname"]} peer {peer_public_key} endpoint to {address}:{port} ... ',
```
```suggestion
                        f'Resetting {self.ifname} peer {peer_public_key} endpoint to {address}:{port} ... ',
```

c-po replied:
Would it be possible to also log the old and new IP addresses for this tunnel?

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -156,6 +156,63 @@ def show_interface(self):
+        from vyos.config import Config
+
+        c = Config()
+        c.set_level(['interfaces', 'wireguard', self.config['ifname']])
```
```suggestion
        c.set_level(['interfaces', 'wireguard', self.ifname])
```

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -187,60 +244,84 @@ def update(self, config):
 
         # Wireguard base command is identical for every peer
-        base_cmd = 'wg set {ifname}'
+        base_cmd = 'wg set ' + config['ifname']
```
```suggestion
        base_cmd = f'wg set {self.ifname}'
```

c-po commented on src/conf_mode/nat.py:
```diff
@@ -17,6 +17,7 @@
 import os
 
 from sys import exit
+from pathlib import Path
```
Why is the move to pathlib required?

sskaje replied:
When there were firewall.py + nat.py, we use /run/use-vyos-domain-resolver-nat and  /run/use-vyos-domain-resolver-firewall, but now we introduced  /run/use-vyos-domain-resolver-wireguard-wgX, we need to change the file-exist-check in all 3 file.

Code in firewall.py is better than nat.py(using not `Path().glob()`), latest: https://github.com/vyos/vyos-1x/blob/ad1f62688ec1ea7319d7aa0bb62082e553be1863/src/conf_mode/firewall.py#L584

Since we need to do same check in 3 files, I changed the nat.py file.

c-po commented on src/helpers/vyos-domain-resolver.py:
```diff
@@ -165,6 +167,47 @@ def update_fqdn(config, node):
+    if node == 'interfaces':
+        wireguard_interfaces = dict_search_args(config, 'wireguard')
+
+        handshake_threshold = 300
```
Please add a comment about this very special number

c-po commented:
Hi @sskaje,

thanks for the PR! I did a testdrive with this implementation and in general it works for me.
Can you please address the open requested changes so we can get this merged?

Please also rebase your work on `current` as its pretty far behind.

sskaje commented:
I'm trying to build and run tests, but I see `frr (>= 10.2)` is now required because of T6746, but the latest [1.5-rolling-202412160007](https://github.com/vyos/vyos-nightly-build/releases/download/1.5-rolling-202412160007/vyos-1.5-rolling-202412160007-generic-amd64.iso) doesn't have frr >= 10.2.

Any Idea how I build and run tests? Revert code in T6746 or just update `debian/control`?

=============

Updated: I modified `debian/control` on my VyOS VM.

c-po commented:
FRR 10.2 is part of the official VyOS repositories now. You do not need to build it on your own.

You could grab a fresh build from e.g. https://github.com/vyos/vyos-nightly-build/actions/runs/12521976266/artifacts/2367044381 

It contains some minor smoketest issues but would be fine for your development. If you're uncertain ant would rather NOT rebase that would work, too - we can rebase in the end after all changes got applied.

c-po commented:
@sskaje you can also force package installation and omit the FRR 10.2 dependency

dpkg --install --force-all *.deb

c-po commented on src/conf_mode/interfaces_wireguard.py:
```diff
@@ -122,6 +130,20 @@ def apply(wireguard):
+    if 'peers_need_resolve' in wireguard and len(wireguard['peers_need_resolve']) > 0:
+        text = f'# Automatically generated by interfaces_wireguard.py\nThis file indicates that vyos-domain-resolver service is used by the interfaces_wireguard.\n'
+        text += "intefaces:\n" + "".join([f"  - {peer}\n" for peer in wireguard['peers_need_resolve']])
+        Path(domain_resolver_usage).write_text(text)
```
Please use our common `vyos.utils.file.write_file()` instead.

Other then that it formally looks good to me - waiting for some more reviewers for this piece of code and another real world test from me.

c-po replied:
```
Dec 29 10:29:43 LR1.wue3 python3[2210]: Resetting wg1000 peer BMj7LgeZS1aoZRQ6rweeKpz+9+HdicT0F590f+7iDH4= endpoint to wg-dynamic.vyos.io:10000 ... done
Dec 29 10:29:43 LR1.wue3 python3[2210]: Wireguard: reset wg1000 peer BMj7LgeZS1aoZRQ6rweeKpz+9+HdicT0F590f+7iDH4=
```

Working

sskaje replied:
Should I perform this file writing change?

c-po replied:
Yes, please do

c-po commented on interface-definitions/interfaces_wireguard.xml.in:
```diff
@@ -98,9 +111,14 @@
                     <format>ipv6</format>
                     <description>IPv6 address of remote tunnel endpoint</description>
                   </valueHelp>
+                  <valueHelp>
```
We might re-thing the CLI node here is it would probably be a better idea to add a new `hostname` CLI node for this particular dyndns case and make it mutually exclusive to `address`

c-po replied:
@sskaje in todays maintainer meeting we discussed the state of this PR and how the FQDN address is passed in.

Can you please perform the following change:

* Remove FQDN support from `address` CLI node
* Add new `host-name` CLI node
* Make `address` and `host-name` mutually exclusive

sskaje replied:
priority address > host-name ?

c-po replied:
They are mutually exclusive, so there is no priority.

sskaje replied:
Where can I find an example for config tree nodes mutally exclusives?

Should I do like this code: https://github.com/vyos/vyos-1x/blob/fe50656a6f1509976d04e8a750083d54bad28ecc/src/conf_mode/high-availability.py#L181

sskaje replied:
Or simple code like 
```

            if 'host_name' in peer and 'address' in peer:
                raise ConfigError('"host-name" and "address" are mutually exclusive')
```

c-po replied:
Use you simple if statement

c-po commented on python/vyos/ifconfig/wireguard.py:
```diff
@@ -156,6 +157,63 @@ def show_interface(self):
+                cmd = f"wg set {self.ifname} peer {peer_public_key} endpoint {address}:{port}"
+                try:
+                    print(
+                        f'Resetting {self.ifname} peer {peer_public_key} endpoint to {address}:{port} ... ',
```
Can you please output the previous endpoint IP address and the new endpoint IP address as well?

c-po commented:
Bug:

When moving from a FQDN address back to an IP address, vyos-domain-resolver is not stopped.

sskaje commented:
> Bug:
> 
> When moving from a FQDN address back to an IP address, vyos-domain-resolver is not stopped.

I tested on my vm, when I move it back to ip address, the /run/use-vyos-domain-resolver-interfaces-wireguard-wgX is gone, and vyos-domain-resolver turn stopped here.

```
-rw-r--r--  1 root     vyattacfg    184 Dec 30 10:17 use-vyos-domain-resolver-interfaces-wireguard-wg0
-rw-r--r--  1 root     vyattacfg    168 Dec 30 10:17 use-vyos-domain-resolver-interfaces-wireguard-wg2
-rw-r--r--  1 root     vyattacfg    182 Dec 30 10:17 use-vyos-domain-resolver-interfaces-wireguard-wg4

```

```
vyos@vyos# set interfaces wireguard wg2 peer wg2-test address 1.1.1.1
[edit]
vyos@vyos# commit
[edit]
vyos@vyos# ls -al /run/|grep use-vyos
-rw-r--r--  1 root     vyattacfg    184 Dec 30 10:17 use-vyos-domain-resolver-interfaces-wireguard-wg0
-rw-r--r--  1 root     vyattacfg    182 Dec 30 10:17 use-vyos-domain-resolver-interfaces-wireguard-wg4
[edit]
vyos@vyos# ps -ef|grep resolver
root       26900       1  0 10:19 ?        00:00:00 /usr/bin/python3 -u /usr/libexec/vyos/vyos-domain-resolver.py
vyos       27026   24131  0 10:19 pts/4    00:00:00 grep resolver
[edit]
vyos@vyos# delete interfaces wireguard wg0
[edit]
vyos@vyos# delete interfaces wireguard wg4
[edit]
vyos@vyos# commit
[edit]
vyos@vyos# ps -ef|grep resolver
vyos       27168   24131  0 10:20 pts/4    00:00:00 grep resolver
[edit]
vyos@vyos# set interfaces wireguard wg2 peer wg2-test address x.x.x.x
[edit]
vyos@vyos# commit
[edit]
vyos@vyos# ps -ef|grep resolver
root       27279       1  4 10:20 ?        00:00:00 /usr/bin/python3 -u /usr/libexec/vyos/vyos-domain-resolver.py
vyos       27350   24131  0 10:20 pts/4    00:00:00 grep resolver
[edit]
vyos@vyos# set interfaces wireguard wg2 peer wg2-test address 1.1.1.1
[edit]
vyos@vyos# commit
[edit]
vyos@vyos# ps -ef|grep resolver
vyos       27512   24131  0 10:20 pts/4    00:00:00 grep resolver
[edit]
```


Can you please check if there is any other `/run/use-vyos-domain-resolver-*` on your side?

c-po commented:
> I tested on my vm, when I move it back to ip address, the /run/use-vyos-domain-resolver-interfaces-wireguard-wgX is gone, and vyos-domain-resolver turn stopped here.

![image](https://github.com/user-attachments/assets/3943cd18-ec95-4782-a1a6-88605a759ab2)

In you example vyos-domain-resolver is started again after you `set interfaces wireguard wg2 peer wg2-test address x.x.x.x` - I think this is invalid.

c-po commented:
This branch has conflicts that must be resolved

* `src/helpers/vyos-domain-resolver.py`

sskaje committed:
wireguard: T4930: allow peers via FQDN

* set interfaces wireguard wgXX peer YY hostname <fqdn>

sskaje commented:
Another thing, 'disable' is defined in wireguard interfaces but not checked when creating interfaces. 

Should I also fix this?

c-po commented:
> Another thing, 'disable' is defined in wireguard interfaces but not checked when creating interfaces.
> 
> Should I also fix this?

This is actually expected behavior. The interface is always created, and `disable` will place it in up/down state.

e.g. `ip link set dev wg up`

This is handle here:
https://github.com/vyos/vyos-1x/blob/99d0c7a804ea3cf7f843f0d4810e6772cf7ceeb8/python/vyos/ifconfig/interface.py#L1832-L1840


I will re-review and test the changes until end of this week.

c-po committed:
smoketest: T4930: derive WireGuard test cases from BasicInterfaceTest

c-po committed:
smoketest: T4930: add wireguard testcase for peer host-name

c-po committed:
wireguard: T4930: use common error message pattern

c-po committed:
wireguard: T4930: add mnemonic for WIREGUARD_REKEY_AFTER_TIME

WireGuard performs a handshake every WIREGUARD_REKEY_AFTER_TIME if data is
being transmitted between the peers. If no data is transmitted, the handshake
will not be initiated unless new data begins to flow. Each handshake generates
a new session key, and the key is rotated at least every 120 seconds or upon
data transmission after a prolonged silence.

c-po commented:
@sskaje I've added a discrete smoketest `test_wireguard_hostname` for the feature and reworked the WireGuard tests in general to benefit from the global interface test framework

```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_interfaces_wireguard.py
test_add_multiple_ip_addresses (__main__.WireGuardInterfaceTest.test_add_multiple_ip_addresses) ... ok
test_add_single_ip_address (__main__.WireGuardInterfaceTest.test_add_single_ip_address) ... ok
test_add_to_invalid_vrf (__main__.WireGuardInterfaceTest.test_add_to_invalid_vrf) ... ok
test_dhcp_client_options (__main__.WireGuardInterfaceTest.test_dhcp_client_options) ... skipped 'not supported'
test_dhcp_disable_interface (__main__.WireGuardInterfaceTest.test_dhcp_disable_interface) ... skipped 'not supported'
test_dhcp_vrf (__main__.WireGuardInterfaceTest.test_dhcp_vrf) ... skipped 'not supported'
test_dhcpv6_client_options (__main__.WireGuardInterfaceTest.test_dhcpv6_client_options) ... skipped 'not supported'
test_dhcpv6_vrf (__main__.WireGuardInterfaceTest.test_dhcpv6_vrf) ... skipped 'not supported'
test_dhcpv6pd_auto_sla_id (__main__.WireGuardInterfaceTest.test_dhcpv6pd_auto_sla_id) ... skipped 'not supported'
test_dhcpv6pd_manual_sla_id (__main__.WireGuardInterfaceTest.test_dhcpv6pd_manual_sla_id) ... skipped 'not supported'
test_eapol (__main__.WireGuardInterfaceTest.test_eapol) ... skipped 'not supported'
test_interface_description (__main__.WireGuardInterfaceTest.test_interface_description) ... ok
test_interface_disable (__main__.WireGuardInterfaceTest.test_interface_disable) ... ok
test_interface_ip_options (__main__.WireGuardInterfaceTest.test_interface_ip_options) ... ok
test_interface_ipv6_options (__main__.WireGuardInterfaceTest.test_interface_ipv6_options) ... ok
test_interface_mtu (__main__.WireGuardInterfaceTest.test_interface_mtu) ... ok
test_ipv6_link_local_address (__main__.WireGuardInterfaceTest.test_ipv6_link_local_address) ... ok
test_move_interface_between_vrf_instances (__main__.WireGuardInterfaceTest.test_move_interface_between_vrf_instances) ... ok
test_mtu_1200_no_ipv6_interface (__main__.WireGuardInterfaceTest.test_mtu_1200_no_ipv6_interface) ... ok
test_span_mirror (__main__.WireGuardInterfaceTest.test_span_mirror) ... skipped 'not supported'
test_vif_8021q_interfaces (__main__.WireGuardInterfaceTest.test_vif_8021q_interfaces) ... skipped 'not supported'
test_vif_8021q_lower_up_down (__main__.WireGuardInterfaceTest.test_vif_8021q_lower_up_down) ... skipped 'not supported'
test_vif_8021q_mtu_limits (__main__.WireGuardInterfaceTest.test_vif_8021q_mtu_limits) ... skipped 'not supported'
test_vif_8021q_qos_change (__main__.WireGuardInterfaceTest.test_vif_8021q_qos_change) ... skipped 'not supported'
test_vif_s_8021ad_vlan_interfaces (__main__.WireGuardInterfaceTest.test_vif_s_8021ad_vlan_interfaces) ... skipped 'not supported'
test_vif_s_protocol_change (__main__.WireGuardInterfaceTest.test_vif_s_protocol_change) ... skipped 'not supported'
test_wireguard_add_remove_peer (__main__.WireGuardInterfaceTest.test_wireguard_add_remove_peer) ... ok
test_wireguard_hostname (__main__.WireGuardInterfaceTest.test_wireguard_hostname) ... ok
test_wireguard_peer (__main__.WireGuardInterfaceTest.test_wireguard_peer) ... ok
test_wireguard_peer_pubkey_change (__main__.WireGuardInterfaceTest.test_wireguard_peer_pubkey_change) ... ok
test_wireguard_same_public_key (__main__.WireGuardInterfaceTest.test_wireguard_same_public_key) ... ok
test_wireguard_threaded (__main__.WireGuardInterfaceTest.test_wireguard_threaded) ... ok

----------------------------------------------------------------------
Ran 32 tests in 270.296s

OK (skipped=15)
```

In addition I did some small refactorings and added a comment about the last handshake/rekey time.

>  WireGuard performs a handshake every WIREGUARD_REKEY_AFTER_TIME if data is
    being transmitted between the peers. If no data is transmitted, the handshake
    will not be initiated unless new data begins to flow. Each handshake generates
    a new session key, and the key is rotated at least every 120 seconds or upon
    data transmission after a prolonged silence.

This behavior raises a small question - is there a possibility to find out when the last byte/message was transmitted?
Imagine a "silent" WireGuard link that will be continously resetted as the handshake was not completed in the given time. Of course, one can enable `persistent-keepalive` on the given peer but that would be another option to enable.

c-po committed:
wireguard: T4930: fix dict key (-/_) when working with config dict

Retrieving the config dict sets key_mangling(), thus we need to look for
host_name instead of host-name.

c-po committed:
Debian: T4930: restart vyos-domain-resolver service if running

c-po committed:
wireguard: T4930: use get_config_dict() rather then individual config queries

Extend ConfigTreeQuery().get_config_dict() with arguments to read in default
CLI values, too. This removes the need for hardcoded default values at
multiple places like:

if max_dns_retry is None:
    max_dns_retry = 3

in this case.

c-po committed:
wireguard: T4930: drop unused WireGuardOperational().show_interface() method

Method is not referenced in the code base, remove dead code.

sskaje commented:
> @sskaje I've added a discrete smoketest `test_wireguard_hostname` for the feature and reworked the WireGuard tests in general to benefit from the global interface test framework
> 
> ```
> vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_interfaces_wireguard.py
> test_add_multiple_ip_addresses (__main__.WireGuardInterfaceTest.test_add_multiple_ip_addresses) ... ok
> test_add_single_ip_address (__main__.WireGuardInterfaceTest.test_add_single_ip_address) ... ok
> test_add_to_invalid_vrf (__main__.WireGuardInterfaceTest.test_add_to_invalid_vrf) ... ok
> test_dhcp_client_options (__main__.WireGuardInterfaceTest.test_dhcp_client_options) ... skipped 'not supported'
> test_dhcp_disable_interface (__main__.WireGuardInterfaceTest.test_dhcp_disable_interface) ... skipped 'not supported'
> test_dhcp_vrf (__main__.WireGuardInterfaceTest.test_dhcp_vrf) ... skipped 'not supported'
> test_dhcpv6_client_options (__main__.WireGuardInterfaceTest.test_dhcpv6_client_options) ... skipped 'not supported'
> test_dhcpv6_vrf (__main__.WireGuardInterfaceTest.test_dhcpv6_vrf) ... skipped 'not supported'
> test_dhcpv6pd_auto_sla_id (__main__.WireGuardInterfaceTest.test_dhcpv6pd_auto_sla_id) ... skipped 'not supported'
> test_dhcpv6pd_manual_sla_id (__main__.WireGuardInterfaceTest.test_dhcpv6pd_manual_sla_id) ... skipped 'not supported'
> test_eapol (__main__.WireGuardInterfaceTest.test_eapol) ... skipped 'not supported'
> test_interface_description (__main__.WireGuardInterfaceTest.test_interface_description) ... ok
> test_interface_disable (__main__.WireGuardInterfaceTest.test_interface_disable) ... ok
> test_interface_ip_options (__main__.WireGuardInterfaceTest.test_interface_ip_options) ... ok
> test_interface_ipv6_options (__main__.WireGuardInterfaceTest.test_interface_ipv6_options) ... ok
> test_interface_mtu (__main__.WireGuardInterfaceTest.test_interface_mtu) ... ok
> test_ipv6_link_local_address (__main__.WireGuardInterfaceTest.test_ipv6_link_local_address) ... ok
> test_move_interface_between_vrf_instances (__main__.WireGuardInterfaceTest.test_move_interface_between_vrf_instances) ... ok
> test_mtu_1200_no_ipv6_interface (__main__.WireGuardInterfaceTest.test_mtu_1200_no_ipv6_interface) ... ok
> test_span_mirror (__main__.WireGuardInterfaceTest.test_span_mirror) ... skipped 'not supported'
> test_vif_8021q_interfaces (__main__.WireGuardInterfaceTest.test_vif_8021q_interfaces) ... skipped 'not supported'
> test_vif_8021q_lower_up_down (__main__.WireGuardInterfaceTest.test_vif_8021q_lower_up_down) ... skipped 'not supported'
> test_vif_8021q_mtu_limits (__main__.WireGuardInterfaceTest.test_vif_8021q_mtu_limits) ... skipped 'not supported'
> test_vif_8021q_qos_change (__main__.WireGuardInterfaceTest.test_vif_8021q_qos_change) ... skipped 'not supported'
> test_vif_s_8021ad_vlan_interfaces (__main__.WireGuardInterfaceTest.test_vif_s_8021ad_vlan_interfaces) ... skipped 'not supported'
> test_vif_s_protocol_change (__main__.WireGuardInterfaceTest.test_vif_s_protocol_change) ... skipped 'not supported'
> test_wireguard_add_remove_peer (__main__.WireGuardInterfaceTest.test_wireguard_add_remove_peer) ... ok
> test_wireguard_hostname (__main__.WireGuardInterfaceTest.test_wireguard_hostname) ... ok
> test_wireguard_peer (__main__.WireGuardInterfaceTest.test_wireguard_peer) ... ok
> test_wireguard_peer_pubkey_change (__main__.WireGuardInterfaceTest.test_wireguard_peer_pubkey_change) ... ok
> test_wireguard_same_public_key (__main__.WireGuardInterfaceTest.test_wireguard_same_public_key) ... ok
> test_wireguard_threaded (__main__.WireGuardInterfaceTest.test_wireguard_threaded) ... ok
> 
> ----------------------------------------------------------------------
> Ran 32 tests in 270.296s
> 
> OK (skipped=15)
> ```
> 
> In addition I did some small refactorings and added a comment about the last handshake/rekey time.
> 
> > WireGuard performs a handshake every WIREGUARD_REKEY_AFTER_TIME if data is
> > being transmitted between the peers. If no data is transmitted, the handshake
> > will not be initiated unless new data begins to flow. Each handshake generates
> > a new session key, and the key is rotated at least every 120 seconds or upon
> > data transmission after a prolonged silence.
> 
> This behavior raises a small question - is there a possibility to find out when the last byte/message was transmitted? Imagine a "silent" WireGuard link that will be continously resetted as the handshake was not completed in the given time. Of course, one can enable `persistent-keepalive` on the given peer but that would be another option to enable.

I don't find anything related to 'when the last byte/message was transmitted' in both kernel and wg-tool, only WGPEER_A_PERSISTENT_KEEPALIVE_INTERVAL.

c-po commented:
> I don't find anything related to 'when the last byte/message was transmitted' in both kernel and wg-tool, only WGPEER_A_PERSISTENT_KEEPALIVE_INTERVAL.

We could add the following logic:

If `(now_time - handshake_time > 3*WIREGUARD_REKEY_AFTER_TIME)` is true and there was no rekey because no data is flowing, we should also check old vs new DNS name and only reset the interface if the resolved DNS name changed.

See `vyos.firewall.fqdn_resolve()`. What you think @sskaje

sskaje commented:
> > I don't find anything related to 'when the last byte/message was transmitted' in both kernel and wg-tool, only WGPEER_A_PERSISTENT_KEEPALIVE_INTERVAL.
> 
> We could add the following logic:
> 
> If `(now_time - handshake_time > 3*WIREGUARD_REKEY_AFTER_TIME)` is true and there was no rekey because no data is flowing, we should also check old vs new DNS name and only reset the interface if the resolved DNS name changed.
> 
> See `vyos.firewall.fqdn_resolve()`. What you think @sskaje


For the rekey time part, I tested on my router,  one of my WG interface got rekey in about every 120-140 seconds. 
3 * rekey time  is enough.

For the DNS part, in fact, resetting peers do reset both address and port of peer endpoint.

I don't think you need to do resolve by python, just let wg do everything, otherwise you need to do domain resolve check, port check, too much work.


For port changing, I have two cases: 

Case A: bad case, one of my friend's network, pppoe, he asked ISP for Internet IP, IP changes when pppoe reconnects. He asked me to set up connection between his network and his cloud server.

His ISP may have some rule like QoS for UDP traffics. I have to periodically change `listen-port` by `wg set wg0 listen-port $PORT` and also keep a fixed port and set up port forwarding to current listen port, so that another side can reset endpoint to its original address:port.  <= port change

His router pppoe gets disconnected randomly every few weeks and Internet IP changes. <= ip change

I connect my router to his network, but I don't need a always online connection, in this case, I need to reset peer if my connection lost.

Case B: similar to case A, port changing policy was also used to "bypass" national firewall's ip+port traffic censorship, but both side have public ip and my ISP behaviors well.
